src/apps/eos/actions/layout.py

```diff
@@ -0,0 +1,405 @@
+    paginator = paginate(show_lines_page, len(pages), FIRST_PAGE, pages, text)
+
+    await ctx.wait(paginator)
+
+
+async def confirm_action_newaccount(ctx, msg: EosActionNewAccount):
+    await ctx.call(
+        ButtonRequest(code=ButtonRequestType.ConfirmOutput), MessageType.ButtonAck
+    )
+
+    text = "New Account"
+    fields = []
+    fields.append("Creator:")
+    fields.append(helpers.eos_name_to_string(msg.creator))
+    fields.append("Name:")
+    fields.append(helpers.eos_name_to_string(msg.name))
+    fields += authorization_fields(msg.owner)
+    fields += authorization_fields(msg.active)
+
+    pages = list(chunks(fields, FOUR_FIELDS_PER_PAGE))
+    paginator = paginate(show_lines_page, len(pages), FIRST_PAGE, pages, text)
+
+    await ctx.wait(paginator)
+
+
+async def confirm_action_unknown(ctx, action, checksum):
+    await ctx.call(
+        ButtonRequest(code=ButtonRequestType.ConfirmOutput), MessageType.ButtonAck
+    )
+    text = "Arbitrary data"
+    fields = []
+    fields.append("Contract:")
+    fields.append(helpers.eos_name_to_string(action.account))
+    fields.append("Action Name:")
+    fields.append(helpers.eos_name_to_string(action.name))
+
+    fields.append("Checksum: ")
+    fields += split_data(hexlify(checksum).decode("ascii"))
+
+    pages = list(chunks(fields, FIVE_FIELDS_PER_PAGE))
+    paginator = paginate(show_lines_page, len(pages), FIRST_PAGE, pages, text)
+
+    await ctx.wait(paginator)
+
+
+@ui.layout
+async def show_lines_page(page: int, page_count: int, pages: list, header: str):
+    if header == "Arbitrary data":
+        text = Text("{} !!!".format(header), ui.ICON_WIPE, icon_color=ui.RED)
+    else:
+        text = Text("{}".format(header), ui.ICON_CONFIRM, icon_color=ui.GREEN)
+    text.mono(*pages[page])
+
+    content = Scrollpage(text, page, page_count)
+    if page + 1 == page_count:
+        if await ConfirmDialog(content) != CONFIRMED:
+            raise wire.ActionCancelled("Action cancelled")
+    else:
+        content.render()
+        await animate_swipe()
+
+
+@ui.layout
+async def show_voter_page(page: int, page_count: int, pages: list):
+    lines = [
+        "%2d. %s" % (wi + 1, helpers.eos_name_to_string(producer))
+        for wi, producer in pages[page]
+    ]
+    text = Text("Vote for producers", ui.ICON_CONFIRM, icon_color=ui.GREEN)
+    text.mono(*lines)
+    content = Scrollpage(text, page, page_count)
+
+    if page + 1 == page_count:
+        if await ConfirmDialog(content) != CONFIRMED:
+            raise wire.ActionCancelled("Action cancelled")
+    else:
+        content.render()
+        await animate_swipe()
+
+
+def authorization_fields(auth):
+    fields = []
+
+    fields.append("Threshold:")
+    fields.append(str(auth.threshold))
+
+    for i, key in enumerate(auth.keys):
+        _key = _public_key_to_wif(bytes(key.key))
+        _weight = str(key.weight)
+
+        header = "Key #{}:".format(i + 1)
```

Comment:
For me `address_n` is awkward feature. Let me explain.
First of all:
> Then afterwards you would derive key from address_n and use that for signing.

Looks like this feature brings more confusion than clearness. The statement is wrong.

EOS accounts have interesting property: permissions. One account may have up to 16 permissions (owner and active are default by design). Basically permissions define which actions with which contracts user can interact with. Each permission might have a key and not necessary the same. On practice it means, user can create a new permission for his account which can interact with action `foo` in contract `bar` and pass it somebody else.
`eosio::updateauth` actions is designed for creating and modifying permissions and by one request user can modify/create up to 16 permissions.

So `address_n` here is just a way to express that modified/new permission will have a key generated by device but with a new path.

Now let's go back awkwardness of the feature:

```
{
    "chain_id": "cf057bbfb72640471fd910bcb67639c22df9f92470936cddc1ade0e2f2e7dc4f",
    "transaction": {
      // ...header
        ...
      "actions": [
        {
          "account": "eosio",
          "name": "updateauth",
          "authorization": [
            {
              "actor": "miniminimini",
              "permission": "active"
            }
          ],
          "data": {
            "account": "miniminimini",
            "permission": "active",
            "parent": "owner",
            "auth": {
              "threshold": 1,
              "keys": [
                {
                  "key": "EOS8.....c6K",
                  "weight": 1
                },
                {
                  "key": "EOS....K",
                  "weight": 2
                }
              ],
              "accounts": [],
              "waits": []
            }
          }
        }
      ],
      "transaction_extensions": []
    }
  }
```
This is a json structure for `eosio::updateauth` action. From this data trezor will build protobuf objects. As you can see here we have two new keys. With existing design it will be transformed directly, signed by device, attached to this json and will be send into the network.

Now what will happen if we will use `address_n` instead of key. First of all in existing eosjs library, tx body is generated automatically by specifying right fields. So we have to generate body by specifying some dummy keys, after modify the structure to add the path and after send it to trezor. Okay now we have the signature. But we don't know what are corresponding public keys to those paths that we used. So for that we have to ask device twice to get public keys for paths that we used previously.

Second case. We are building json without eosjs library, manually, with paths. Send it to trezor. Sign the transaction. But still we have to know public keys otherwise transaction will be rejected. Again, we have to ask device to get public keys.

In case of 16 permissions, we have to ask the device 16 times. Really?

Even if it will be implemented I don't think clients/wallets will support it. 
Do you see any reasons to include it?

Reply:
`address_n` here is an annotation that the device can use to give the user "richer" information about what they're doing, reducing the chances that they'll screw up, and confirm something that they didn't want to. Clients that implement it will be giving their users a safer UX, which is always something we should be striving for in wallet implementations.

Reply:
An analogous thing on Bitcoin tx's is change addresses... when was the last time you saw a wallet that asked you to confirm precisely where change was going?

Reply:
@tarrash if you're using public keys from Trezor accounts, you need to send the GetPublicKey queries anyway. Also these are silent unless you specify `show_display`, no?

The way I see it, the client-side algorithm is the following.
Without address_n:
1. request public key from device
2. generate tx body, using this public key
3. fill out protobufs with this public key
4. request signature from device

with address_n:
1. request public key from device
2. generate tx body, using this public key
3. fill out protobufs *with empty pubkey and appropriate address_n*
4. request signature from device

I agree that the implementation of step (3) is awkward, it requires you to edit auto-translated protobufs. But it allows you to add permissions for your own accounts safely without requiring the user to write down their own pubkey and then compare it to the screen.
Which seems like an important UX convenience for this particular scenario - something you really want, if the scenario is at least somewhat realistic.

I see this as an additional feature on top of the existing interface, so it's technically optional for this PR. It's a -0 from me for leaving it out, but if other reviewers agree, we can merge without it.

Reply:
Guys, it seems that's the last open item. So what do we do, can we resolve it and leave that feature out? We can consider it as a future improvement and add somewhere in the future releases.

Reply:
OK, let's add it later.